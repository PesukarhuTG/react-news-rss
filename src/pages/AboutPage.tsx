import React from 'react';
import styled from 'styled-components';
import { Layout } from '../components';

const AboutPage: React.FC = () => {
  return (
    <Layout>
      <Wrapper>
        <Title>About us</Title>
        <Description>
          <b>Racoon news</b>: it is the latest breaking news and information on the top stories,
          weather, business, entertainment, politics, and more.
        </Description>
        <PrimaryText>You do not miss anything!</PrimaryText>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Description = styled.p`
  margin-top: 30px;
  text-align: center;
  font-size: 18px;
`;

const PrimaryText = styled.p`
  margin-top: 30px;
  text-align: center;
  font-size: 22px;
  font-weight: 700;
`;

export default AboutPage;
