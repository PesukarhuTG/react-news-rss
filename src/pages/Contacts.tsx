import React from 'react';
import styled from 'styled-components';
import FormProps from '../types/Form';
import { Layout, LoginForm, FormCardsAlbum } from '../components';

interface State {
  list?: FormProps[];
  message?: string;
}

class Contacts extends React.Component<State> {
  state = {
    list: [],
    message: '',
  };

  onSubmit = (formFields: FormProps) => {
    const currentList = this.state.list as FormProps[];
    currentList.push(formFields);
    this.setState({ list: currentList });

    this.setState({ message: 'Your data has been saved!' });
    setTimeout(() => this.setState({ message: '' }), 1500);
  };

  render() {
    return (
      <Layout>
        <Wrapper>
          <Title>Contact us via form</Title>
          <LoginForm onSubmit={this.onSubmit} />
        </Wrapper>
        <Message data-testid="form-message">{this.state.message}</Message>
        <FormCardsAlbum list={this.state.list} />
      </Layout>
    );
  }
}

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Message = styled.p`
  height: 30px;
  font-size: 14px;
  font-weight: 700;
  text-align: center;
  color: var(--accent);
`;

export default Contacts;
