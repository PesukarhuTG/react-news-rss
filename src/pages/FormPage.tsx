import React, { useState } from 'react';
import styled from 'styled-components';
import FormProps from '../types/Form';
import { Layout, Form, FormCardsAlbum } from '../components';
import { useDispatch, useSelector } from 'react-redux';
import { changeFormList } from 'store/FormSlice';
import { AppDispatch, RootState } from 'store/Store';

const FormPage: React.FC = () => {
  const [message, setMessage] = useState<string>('');
  const { formList } = useSelector((state: RootState) => state.forms);
  const dispatch = useDispatch<AppDispatch>();

  const onSubmit = (formFields: FormProps) => {
    const currentList = [...formList];
    currentList.push(formFields);
    dispatch(changeFormList(currentList));

    setMessage('Your data has been saved!');
    setTimeout(() => setMessage(''), 1500);
  };

  return (
    <Layout>
      <Wrapper>
        <Title>Save your data</Title>
        <Form onSubmit={onSubmit} />
      </Wrapper>
      <Message data-testid="form-message">{message}</Message>
      <FormCardsAlbum list={formList} />
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Message = styled.p`
  height: 30px;
  font-size: 14px;
  font-weight: 700;
  text-align: center;
  color: var(--accent);
`;

export default FormPage;
