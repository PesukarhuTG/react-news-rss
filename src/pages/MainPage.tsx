import React, { useState, useEffect } from 'react';
import {
  Layout,
  SearchPanel,
  CardsAlbum,
  SortSelectBy,
  SortSelectIn,
  SortDateFrom,
  SortDateTo,
  Pagination,
} from 'components';
import styled from 'styled-components';
import { getNews, searchNews } from '../services/getDataApi';
import CardProps from '../types/Card';
import useNewsContext from '../store/Context';

const MainPage: React.FC = () => {
  const [news, setNews] = useState<CardProps[]>([]);
  const [message, setMessage] = useState<string>('');

  const {
    searchVal,
    searchIn,
    sortBy,
    sortDateFrom,
    sortDateTo,
    currentPage,
    setCurrentPage,
    totalPageAmount,
    setTotalPageAmount,
    pageSize,
    setPageSize,
  } = useNewsContext();

  const onSubmit = (data: CardProps[]): void => {
    if (!data.length) {
      setMessage('Sorry, your request is failed');
    } else {
      setMessage('');
    }
    setNews(data);
  };

  useEffect(() => {
    const checkData = async (): Promise<void> => {
      if (searchVal) {
        const data = await searchNews(
          searchVal,
          searchIn,
          sortBy,
          sortDateFrom,
          sortDateTo,
          currentPage,
          pageSize
        ).then((resp) => resp);

        if (!data.articles.length) {
          setMessage('Sorry, your request is failed');
        } else {
          setMessage('');
          setNews(data.articles);
          setTotalPageAmount(data.totalResults > 100 ? 100 : data.totalResults);
        }
      } else {
        const data = await getNews(currentPage, pageSize).then((resp) => resp);
        setNews(data.articles);
        setTotalPageAmount(data.totalResults > 100 ? 100 : data.totalResults);
      }
    };

    checkData();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchIn, sortBy, sortDateFrom, sortDateTo, currentPage, pageSize]);

  return (
    <Layout>
      <SearchPanel onSearch={onSubmit} />
      <Headling>Hot news on Racoon digest</Headling>
      <SortWrapper>
        <SortBlock>
          <p>Searching by:</p>
          <SortSelectIn options={['title', 'description']} />
          <SortSelectBy options={['publishedAt', 'relevancy', 'popularity']} />
        </SortBlock>
        <SortBlock>
          <SortDateFrom />
          <SortDateTo />
        </SortBlock>
      </SortWrapper>
      <Message data-testid="fail-message">{message}</Message>
      <CardsAlbum cards={news} />
      <Pagination
        page={currentPage}
        onChange={(page, pageSize) => {
          setCurrentPage(page);
          setPageSize(pageSize);
        }}
        total={totalPageAmount}
        pageSize={pageSize}
      />
    </Layout>
  );
};

const Headling = styled.p`
  margin: 20px 0 10px;
  text-align: center;
  font-size: 36px;
  font-weight: 700;
  text-transform: uppercase;
  color: var(--primary);
  opacity: 0.6;
`;

const Message = styled.p`
  margin: 10px 0 10px;
  height: 20px;
  text-align: center;
  font-size: 16px;
  font-weight: 700;
  color: var(--accent);
`;

const SortBlock = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 20px;
`;

const SortWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 20px;
  gap: 20px;
  background-color: var(--second-contrast);
  font-size: 14px;

  @media (max-width: 950px) {
    flex-direction: column;
  }
`;

export default MainPage;
