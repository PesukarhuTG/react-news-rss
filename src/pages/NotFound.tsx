import React from 'react';
import styled from 'styled-components';
import { Layout } from '../components';
import { Link } from 'react-router-dom';

const NotFound: React.FC = () => {
  return (
    <Layout>
      <Wrapper>
        <Title>404</Title>
        <Description>sorry, page not found</Description>
        <HomeLink to="/">â–º back home</HomeLink>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 140px;
  font-weight: 700;
`;

const Description = styled.p`
  text-align: center;
  font-size: 24px;
  letter-spacing: 1.5px;
  margin: 0 0 20px;
`;

const HomeLink = styled(Link)`
  display: block;
  text-align: center;
  color: var(--primary);

  &:hover {
    text-decoration: underline;
  }
`;

export default NotFound;
