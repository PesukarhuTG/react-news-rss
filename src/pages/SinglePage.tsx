import React, { useEffect } from 'react';
import styled from 'styled-components';
import { Layout } from '../components';
import { useNavigate } from 'react-router-dom';
import useNewsContext from '../store/Context';
import SavedCardProps from '../types/SavedCardData';

const SinglePage: React.FC = () => {
  const navigate = useNavigate();
  const { savedCardData, setDisableCurrentPosition } = useNewsContext();

  const { author, description, publishedAt, title, urlToImage, url } =
    savedCardData as SavedCardProps;

  const backMainPage = () => {
    navigate('/', { replace: true });
  };

  useEffect(() => {
    if (!author && !description && !publishedAt && !title && !urlToImage && !url) {
      navigate('/', { replace: true });
    }

    return () => {
      setDisableCurrentPosition(true);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Layout>
      <Wrapper>
        <NewsImage
          style={{
            backgroundImage: `url(${urlToImage}`,
          }}
        />
        <NewsFullTitle>{title}</NewsFullTitle>

        <NewsFullDescription>
          {description || 'Sorry, this news has no description'}
        </NewsFullDescription>
        <InfoWrapper>
          <PublishedDiv>
            <span>
              <b>Published date</b>
            </span>
            <span>{publishedAt ? publishedAt.slice(0, 10) : 'no date'}</span>
          </PublishedDiv>
          <PublishedDiv>
            <span>
              <b>Author</b>
            </span>
            <span>{author || 'incognita'}</span>
          </PublishedDiv>
        </InfoWrapper>
        <InfoWrapper>
          <Button onClick={backMainPage}>❮ Back to main page</Button>
          <a href={url} target="_blank" rel="noreferrer noopener">
            <Button>Link to full news ❯</Button>
          </a>
        </InfoWrapper>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  max-width: 500px;
  width: 100%;
  margin: 0 auto;
`;

const InfoWrapper = styled.div`
  width: 100%;
  display: flex;
  font-size: 14px;
  gap: 20px;
  margin: 20px 0;
  align-items: center;
  justify-content: space-between;
`;

const NewsImage = styled.div`
  width: 100%;
  height: 240px;
  margin: 40px 0 30px;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  border-radius: 10px;
  box-shadow: 0 0 15px 0 var(--primary);
`;

const NewsFullTitle = styled.p`
  margin-bottom: 20px;
  text-align: center;
  font-size: 18px;
  font-weight: 700;
  color: var(--primary);
  text-transform: uppercase;
`;

const NewsFullDescription = styled.p`
  font-size: 14px;
  padding-bottom: 20px;
  border-bottom: 1px dotted var(--main-color);
`;

const Button = styled.button`
  display: block;
  max-width: 170px;
  width: 100%;
  padding: 10px;
  font-size: 14px;
  border-radius: 5px;
  background-color: var(--primary);
  border: 0;
  cursor: pointer;
  transition: 0.3s;
  &:hover {
    background-color: var(--button-hover);
  }
`;

const PublishedDiv = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

export default SinglePage;
