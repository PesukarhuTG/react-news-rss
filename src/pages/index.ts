export { default as MainPage } from './MainPage';
export { default as AboutPage } from './AboutPage';
export { default as NotFound } from './NotFound';
export { default as FormPage } from './FormPage';
export { default as SinglePage } from './SinglePage';
