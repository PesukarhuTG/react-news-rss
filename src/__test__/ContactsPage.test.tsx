import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import { Contacts } from '../pages';

describe('Contacts page tests', () => {
  test('Contacts page render', () => {
    render(
      <BrowserRouter>
        <Contacts />
      </BrowserRouter>
    );

    const title = screen.getByText(/Contact us via form/i);
    expect(title).toBeInTheDocument();
  });

  test('Contacts page snapshot', () => {
    const page = render(
      <BrowserRouter>
        <Contacts />
      </BrowserRouter>
    );
    expect(page).toMatchSnapshot();
  });

  test('render Form on Contacts page', () => {
    render(
      <BrowserRouter>
        <Contacts />
      </BrowserRouter>
    );

    expect(screen.getByTestId('form')).toBeInTheDocument();
  });
});
