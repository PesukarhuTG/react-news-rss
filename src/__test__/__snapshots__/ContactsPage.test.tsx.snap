// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts page tests Contacts page snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkzDqf dbxxpE"
        data-testid="layout-container"
      >
        <header
          class="sc-bczRLJ fDDTxG"
        >
          <a
            aria-current="page"
            class="logo active"
            href="/"
          >
            RACOON digest
          </a>
          <nav
            class="sc-gsnTZi fwfCFK"
          >
            <a
              aria-current="page"
              class="active-link"
              data-testid="mainpage-link"
              href="/"
            >
              Home
            </a>
            <a
              data-testid="aboutpage-link"
              href="/about"
            >
              About
            </a>
            <a
              data-testid="contactspage-link"
              href="/contacts"
            >
              Contacts
            </a>
          </nav>
        </header>
        <div
          class="sc-llJcti gcQQGp"
        >
          <h2
            class="sc-iIPllB eKoJl"
          >
            Contact us via form
          </h2>
          <form
            data-testid="form"
          >
            <div
              class="form-wrapper"
            >
              <label
                class="input-wrapper"
              >
                <span
                  class="input-name"
                >
                  Name
                </span>
                <input
                  class="input-field"
                  data-testid="input-fname"
                  name="name"
                  required=""
                  type="text"
                />
              </label>
              <label
                class="input-wrapper"
              >
                <span
                  class="input-name"
                >
                  Birthday
                </span>
                <input
                  class="input-field"
                  data-testid="input-fdata"
                  name="birthday"
                  required=""
                  type="date"
                />
              </label>
            </div>
            <div
              class="form-wrapper"
            >
              <label
                class="input-wrapper"
              >
                <span
                  class="input-name"
                >
                  Choose a city
                </span>
                <select
                  class="input-city"
                  data-testid="input-fcity"
                  name="city"
                  required=""
                >
                  <option
                    value=""
                  />
                  <option
                    value="Saint-Petersburg"
                  >
                    Saint-Petersburg
                  </option>
                  <option
                    value="Moscow"
                  >
                    Moscow
                  </option>
                  <option
                    value="Voronezh"
                  >
                    Voronezh
                  </option>
                  <option
                    value="Petrozavodsk"
                  >
                    Petrozavodsk
                  </option>
                  <option
                    value="other city"
                  >
                    ...other city
                  </option>
                </select>
              </label>
              <label
                class="input-wrapper"
              >
                <span
                  class="input-name"
                >
                  Gender
                </span>
                <div
                  class="genders-wrapper"
                >
                  <input
                    checked=""
                    data-testid="input-fgender-man"
                    id="man"
                    name="gender"
                    type="radio"
                    value="man"
                  />
                  <label
                    class="gender-name"
                    for="man"
                  >
                    man
                  </label>
                  <input
                    data-testid="input-fgender-woman"
                    id="woman"
                    name="gender"
                    type="radio"
                    value="woman"
                  />
                  <label
                    class="gender-name"
                    for="woman"
                  >
                    woman
                  </label>
                </div>
              </label>
            </div>
            <label
              class="input-wrapper"
            >
              <span
                class="input-name"
              >
                Upload a file
              </span>
              <input
                accept="image/*, .png, .jpg"
                class="input-file"
                data-testid="input-file"
                name="file"
                type="file"
              />
            </label>
            <label
              class="input-wrapper-gorizont"
            >
              <input
                data-testid="input-faccept"
                name="remember"
                type="checkbox"
              />
              <span
                class="remember-name"
              >
                I consent to use my personal data
              </span>
            </label>
            <button
              class="btn-submit"
              data-testid="btn-submit"
              disabled=""
              type="submit"
            >
              Send my data
            </button>
          </form>
        </div>
        <p
          class="sc-gicCDI hyttfp"
          data-testid="form-message"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkzDqf dbxxpE"
      data-testid="layout-container"
    >
      <header
        class="sc-bczRLJ fDDTxG"
      >
        <a
          aria-current="page"
          class="logo active"
          href="/"
        >
          RACOON digest
        </a>
        <nav
          class="sc-gsnTZi fwfCFK"
        >
          <a
            aria-current="page"
            class="active-link"
            data-testid="mainpage-link"
            href="/"
          >
            Home
          </a>
          <a
            data-testid="aboutpage-link"
            href="/about"
          >
            About
          </a>
          <a
            data-testid="contactspage-link"
            href="/contacts"
          >
            Contacts
          </a>
        </nav>
      </header>
      <div
        class="sc-llJcti gcQQGp"
      >
        <h2
          class="sc-iIPllB eKoJl"
        >
          Contact us via form
        </h2>
        <form
          data-testid="form"
        >
          <div
            class="form-wrapper"
          >
            <label
              class="input-wrapper"
            >
              <span
                class="input-name"
              >
                Name
              </span>
              <input
                class="input-field"
                data-testid="input-fname"
                name="name"
                required=""
                type="text"
              />
            </label>
            <label
              class="input-wrapper"
            >
              <span
                class="input-name"
              >
                Birthday
              </span>
              <input
                class="input-field"
                data-testid="input-fdata"
                name="birthday"
                required=""
                type="date"
              />
            </label>
          </div>
          <div
            class="form-wrapper"
          >
            <label
              class="input-wrapper"
            >
              <span
                class="input-name"
              >
                Choose a city
              </span>
              <select
                class="input-city"
                data-testid="input-fcity"
                name="city"
                required=""
              >
                <option
                  value=""
                />
                <option
                  value="Saint-Petersburg"
                >
                  Saint-Petersburg
                </option>
                <option
                  value="Moscow"
                >
                  Moscow
                </option>
                <option
                  value="Voronezh"
                >
                  Voronezh
                </option>
                <option
                  value="Petrozavodsk"
                >
                  Petrozavodsk
                </option>
                <option
                  value="other city"
                >
                  ...other city
                </option>
              </select>
            </label>
            <label
              class="input-wrapper"
            >
              <span
                class="input-name"
              >
                Gender
              </span>
              <div
                class="genders-wrapper"
              >
                <input
                  checked=""
                  data-testid="input-fgender-man"
                  id="man"
                  name="gender"
                  type="radio"
                  value="man"
                />
                <label
                  class="gender-name"
                  for="man"
                >
                  man
                </label>
                <input
                  data-testid="input-fgender-woman"
                  id="woman"
                  name="gender"
                  type="radio"
                  value="woman"
                />
                <label
                  class="gender-name"
                  for="woman"
                >
                  woman
                </label>
              </div>
            </label>
          </div>
          <label
            class="input-wrapper"
          >
            <span
              class="input-name"
            >
              Upload a file
            </span>
            <input
              accept="image/*, .png, .jpg"
              class="input-file"
              data-testid="input-file"
              name="file"
              type="file"
            />
          </label>
          <label
            class="input-wrapper-gorizont"
          >
            <input
              data-testid="input-faccept"
              name="remember"
              type="checkbox"
            />
            <span
              class="remember-name"
            >
              I consent to use my personal data
            </span>
          </label>
          <button
            class="btn-submit"
            data-testid="btn-submit"
            disabled=""
            type="submit"
          >
            Send my data
          </button>
        </form>
      </div>
      <p
        class="sc-gicCDI hyttfp"
        data-testid="form-message"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
